### PR TITLE
[SPARK-31456][CORE] Fix shutdown hook priority edge cases

### PR DESCRIPTION
**Pull request by oleg-smith**
### What changes were proposed in this pull request?
Fix application order for shutdown hooks for the priorities of Int.MaxValue, Int.MinValue

### Why are the changes needed?
The bug causes out-of-order execution of shutdown hooks if their priorities were Int.MinValue or Int.MaxValue

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added a test covering the change.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 1]
```diff
@@ -209,9 +209,7 @@ private [util] class SparkShutdownHookManager {
 private class SparkShutdownHook(private val priority: Int, hook: () => Unit)
   extends Comparable[SparkShutdownHook] {
 
-  override def compareTo(other: SparkShutdownHook): Int = {
-    other.priority - priority
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala:213]: Do you mean the overflow cause out of order?

But isn't `SparkShutdownHookManager` is an internal class? Where do we use `MaxValue`/`MinValue` and cause the problem?

Comment (oleg-smith) [reply core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala:213]: yes, it caused it. I'm not sure about the usage.

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 2]
```diff
@@ -209,9 +209,7 @@ private [util] class SparkShutdownHookManager {
...
-  }
+  override def compareTo(other: SparkShutdownHook): Int = Integer.compare(other.priority, priority)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala:212]: why don't you just call `other.priority.compareTo(priority)`?

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala (hunk 1 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala:213]: Yeah, I don't think this affects any external case.

**Comment by HyukjinKwon** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala (hunk 2 above)
Comment (oleg-smith) [reply core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala:212]: no reason

**Comment by HyukjinKwon** [comment]
I think we can just treat this target as a cleanup.

**Comment by oleg-smith** [comment]
@HyukjinKwon ok to merge?

**Review by srowen — APPROVED** [review]
That's a fine clean fix, regardless of whether it happens in real usage now.

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala (hunk 2 above)
Comment (srowen) [reply core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: Either is fine. Integer.compare() avoids boxing but it won't matter at all here

**Review by HyukjinKwon — APPROVED** [review]

**Comment by Ngone51** [comment]
LGTM

**Comment by oleg-smith** [comment]
@srowen @HyukjinKwon ok to merge?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you all
Merged to master/3.0/2.4.

**Comment by dongjoon-hyun** [comment]
Thank you for your first contribution, @oleg-smith .
You are added to the Apache Spark contributor group and SPARK-31456 is assigned to you. :)

**Pull request closed without merging**